hathor/transaction/storage/transaction_storage.py

```diff
@@ -333,6 +344,36 @@ def get_height_best_block(self) -> int:
 
         return highest_height
 
+    def get_merkle_tree(self, timestamp: int) -> Tuple[bytes, List[bytes]]:
```

Comment:
Even though I got the point, I'm not sure we should move both methods to the storage because they are exclusively used by the sync algorithm. Can't we do the cache inside the sync code?

Reply:
We probably have to cache on p2p manager, as we have one node_sync instance per peer.

Reply:
Yes, my first implementation was on the p2p manager. However it was not easy to invalidate the cache and I had to use the pubsub to get the events and invalidate the cache.

We already know that we have a bug where we emit repeated pubsub events, so I was invalidating the cache in some moments I shouldn't, what could make the node waste CPU.

That's why I moved to the storage to use the same method to invalidate the cache as the one from the tips.

If you have any thoughts about how to improve this, I can refactor the code.

Reply:
Why was it needed to use pubsub? `NodeSync` has access to `HathorManager`, which in turn can access `ConnectionsManager`. But that seems like hack. 

A better approach could be adding `ConnectionsManager` to `NodeSync`. When we create `NodeSync`, we have access to the protocol, which stores this object. We could add it also there.
```
    def __init__(self, protocol: 'HathorProtocol', reactor: Clock = None) -> None:
        ...
        self.protocol = protocol
        self.manager = protocol.node
        # something like this
        self.connections = protocol.connections
```

Reply:
Why do I need `ConnectionsManager`? I need every moment a tx is added to the cache.

Reply:
`ConnectionsManager` is the p2p manager. I thought this would be the best place to have the cache. It can be accessed by the `NodeSync` code (if you make the changes above). It can also be accessed on the storage through `HathorManager`.

It may be easier to have the cache in `HathorManager`, but since it's related to sync, having it inside `hathor/p2p/` seems more logical.

Reply:
Ah, okay, now I understand. So add the cache to the `ConnectionsManager` and make the storage call a `reset_cache` method on it every time a tx is added to the cache?

To be honest I prefer to keep all caches on the storage but I can do this refactor if you feel is better.

@msbrogli @obiyankenobi

Reply:
Yeah, that's the idea.

Regarding the location, I was looking at all the caches in our storage and it looks a bit confusing. We'll probably need to refactor it at some point. So for this index, I'd vote for adding it under the sync code, but I'm also fine if we add it to the storage with all the other indexes.

Reply:
I finally understood the difficulty to put the cache out of the storage. I agree that it can be in the storage for now. Even though I agree with @obiyankenobi that we'll need to refactor it later, I don't think we should put it in the `ConnectionsManager`.

Reply:
Ok, so will keep on the storage for now.